pkg/crds/crds_test.go

```diff
@@ -0,0 +1,174 @@
+// Package crds is used for installing rancher CRDs
+package crds
+
+import (
+	"context"
+	"embed"
+	"sort"
+	"testing"
+
+	"github.com/rancher/rancher/pkg/features"
+	"github.com/rancher/rancher/pkg/fleet"
+	"github.com/stretchr/testify/require"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	fakeclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1/fake"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8stesting "k8s.io/client-go/testing"
+)
+
+const (
+	capiCRD           = "clusters.cluster.x-k8s.io"
+	rtCRD             = "roletemplates.management.cattle.io"
+	bootstrapFleetCRD = "clusters.fleet.cattle.io"
+)
+
+var staticCRD = &apiextv1.CustomResourceDefinition{
+	TypeMeta:   v1.TypeMeta{Kind: "CustomResourceDefinition", APIVersion: "apiextensions.k8s.io/v1"},
+	ObjectMeta: v1.ObjectMeta{Name: "crdName"},
+	Status: apiextv1.CustomResourceDefinitionStatus{Conditions: []apiextv1.CustomResourceDefinitionCondition{
+		{
+			Type:    "NamesAccepted",
+			Status:  "True",
+			Reason:  "NoConflicts",
+			Message: "no conflicts found",
+		},
+		{
+			Type:    "Established",
+			Status:  "True",
+			Reason:  "InitialNamesAccepted",
+			Message: "the initial names have been accepted",
+		},
+	}},
+}
+
+func TestEnsure_MCM(t *testing.T) {
```

Comment:
Nit:

There are a couple of test cases that this doesn't cover, namely:
- Failure to create the CRDs in k8s
- Duplicate CRDs (might be tough to do this one, you'd probably need to write some fake yaml to a file)
- Malformed Yaml (i.e. can't unmarshal file to yaml)
- File with content that is marshallable to a CRD but isn't of a Kind "CustomResourceDefinition"

Reply:
Added Test case for each of these areas.

Reply:
Thanks for adding those tests.